File: go/internal/query/optimizer.go
package distributed

import (
	"math"
	"sort"

	stor "github.com/knirv/knirvbase/internal/storage"
)

// QueryPlan represents an optimized execution plan for a query
type QueryPlan struct {
	// Execution strategy
	UseIndex  bool
	IndexName string
	IndexType stor.IndexType
	ScanType  ScanType // FullScan, IndexScan, IndexOnlyScan

	// Query parameters
	Filters   []Filter
	SortOrder SortOrder
	Limit     int

	// Cost estimation
	EstimatedCost float64
	EstimatedRows int

	// Execution details
	IndexFilters []Filter // Filters that can be pushed to index
	PostFilters  []Filter // Filters that need post-processing
}

// ScanType represents the type of scan to perform
type ScanType int

const (
	FullScan ScanType = iota
	IndexScan
	IndexOnlyScan
)

// SortOrder represents sorting requirements
type SortOrder struct {
	Field     string
	Ascending bool
}

// QueryOptimizer analyzes queries and creates optimal execution plans
type QueryOptimizer struct {
	collection string
	indexes    []*stor.Index
	stats      *CollectionStats
}

// CollectionStats holds statistics about a collection for cost estimation
type CollectionStats struct {
	TotalDocuments int
	IndexStats     map[string]*IndexStats
}

// IndexStats holds statistics about an index
type IndexStats struct {
	Cardinality   int     // Number of unique values
	Selectivity   float64 // Fraction of documents matching a filter (0-1)
	AvgBucketSize float64 // Average number of documents per index key
}

// NewQueryOptimizer creates a new query optimizer
func NewQueryOptimizer(collection string, indexes []*stor.Index, stats *CollectionStats) *QueryOptimizer {
	if stats == nil {
		stats = &CollectionStats{
			TotalDocuments: 1000, // Default estimate
			IndexStats:     make(map[string]*IndexStats),
		}
		// Initialize default stats for indexes
		for _, idx := range indexes {
			stats.IndexStats[idx.Name] = &IndexStats{
				Cardinality:   100,
				Selectivity:   0.1,
				AvgBucketSize: 10.0,
			}
		}
	}

	return &QueryOptimizer{
		collection: collection,
		indexes:    indexes,
		stats:      stats,
	}
}

// Optimize creates an optimal query plan for the given query
func (qo *QueryOptimizer) Optimize(query *Query) (*QueryPlan, error) {
	plan := &QueryPlan{
		Filters:       query.Filters,
		SortOrder:     SortOrder{}, // Not implemented yet
		Limit:         query.Limit,
		ScanType:      FullScan,
		EstimatedRows: qo.stats.TotalDocuments,
	}

	// If no filters, always do full scan
	if len(query.Filters) == 0 {
		plan.EstimatedCost = float64(qo.stats.TotalDocuments)
		return plan, nil
	}

	// Analyze filters and find best index
	bestIndex, indexFilters, postFilters := qo.selectBestIndex(query.Filters)

	if bestIndex != nil {
		plan.UseIndex = true
		plan.IndexName = bestIndex.Name
		plan.IndexType = bestIndex.Type
		plan.IndexFilters = indexFilters
		plan.PostFilters = postFilters

		// Estimate cost and rows for index scan
		plan.EstimatedRows = qo.estimateIndexRows(bestIndex, indexFilters)
		plan.EstimatedCost = qo.estimateIndexCost(bestIndex, indexFilters, plan.EstimatedRows)

		// Determine scan type
		if len(postFilters) == 0 {
			plan.ScanType = IndexOnlyScan
		} else {
			plan.ScanType = IndexScan
		}
	} else {
		// Full scan
		plan.EstimatedCost = float64(qo.stats.TotalDocuments)
		plan.PostFilters = query.Filters
	}

	return plan, nil
}

// selectBestIndex chooses the best index for the given filters
func (qo *QueryOptimizer) selectBestIndex(filters []Filter) (*stor.Index, []Filter, []Filter) {
	if len(filters) == 0 {
		return nil, nil, nil
	}

	type indexCandidate struct {
		index        *stor.Index
		indexFilters []Filter
		postFilters  []Filter
		cost         float64
		rows         int
	}

	var candidates []indexCandidate

	// Evaluate each index
	for _, idx := range qo.indexes {
		indexFilters, postFilters := qo.analyzeIndexSuitability(idx, filters)

		if len(indexFilters) > 0 {
			estimatedRows := qo.estimateIndexRows(idx, indexFilters)
			estimatedCost := qo.estimateIndexCost(idx, indexFilters, estimatedRows)

			candidates = append(candidates, indexCandidate{
				index:        idx,
				indexFilters: indexFilters,
				postFilters:  postFilters,
				cost:         estimatedCost,
				rows:         estimatedRows,
			})
		}
	}

	if len(candidates) == 0 {
		return nil, nil, filters
	}

	// Sort by cost (lowest first)
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].cost < candidates[j].cost
	})

	best := candidates[0]
	return best.index, best.indexFilters, best.postFilters
}

// analyzeIndexSuitability determines which filters can use an index
func (qo *QueryOptimizer) analyzeIndexSuitability(idx *stor.Index, filters []Filter) ([]Filter, []Filter) {
	var indexFilters []Filter
	var postFilters []Filter

	for _, filter := range filters {
		if qo.canUseIndex(idx, filter) {
			indexFilters = append(indexFilters, filter)
		} else {
			postFilters = append(postFilters, filter)
		}
	}

	return indexFilters, postFilters
}

// canUseIndex checks if a filter can use the given index
func (qo *QueryOptimizer) canUseIndex(idx *stor.Index, filter Filter) bool {
	switch idx.Type {
	case stor.IndexTypeBTree:
		// B-Tree indexes support exact matches and range queries
		return qo.canUseBTreeIndex(idx, filter)
	case stor.IndexTypeGIN:
		// GIN indexes support text search and JSON queries
		return qo.canUseGINIndex(idx, filter)
	case stor.IndexTypeHNSW:
		// HNSW indexes are for vector similarity
		return false // Not for regular filters
	default:
		return false
	}
}

// canUseBTreeIndex checks if a B-Tree index can be used for a filter
func (qo *QueryOptimizer) canUseBTreeIndex(idx *stor.Index, filter Filter) bool {
	// Check if the filter field is indexed
	for _, field := range idx.Fields {
		if filter.Key == field {
			return true
		}
	}
	return false
}

// canUseGINIndex checks if a GIN index can be used for a filter
func (qo *QueryOptimizer) canUseGINIndex(_ *stor.Index, _ Filter) bool {
	// GIN indexes can be used for text search
	// For now, support CONTAINS operations
	return true // Simplified - assume GIN can handle most text filters
}

// estimateIndexRows estimates the number of rows returned by an index scan
func (qo *QueryOptimizer) estimateIndexRows(idx *stor.Index, filters []Filter) int {
	if len(filters) == 0 {
		return qo.stats.TotalDocuments
	}

	stats := qo.stats.IndexStats[idx.Name]
	if stats == nil {
		// Default estimate
		return int(float64(qo.stats.TotalDocuments) * 0.1)
	}

	// Start with total documents
	estimatedRows := float64(qo.stats.TotalDocuments)

	// Apply selectivity for each filter
	for _, filter := range filters {
		selectivity := qo.estimateFilterSelectivity(idx, filter)
		estimatedRows *= selectivity
	}

	return int(math.Max(1, estimatedRows))
}

// estimateFilterSelectivity estimates the selectivity of a filter
func (qo *QueryOptimizer) estimateFilterSelectivity(idx *stor.Index, _ Filter) float64 {
	stats := qo.stats.IndexStats[idx.Name]
	if stats == nil {
		return 0.1 // Default selectivity
	}

	// For B-Tree indexes, use cardinality
	if idx.Type == stor.IndexTypeBTree {
		if stats.Cardinality > 0 {
			return 1.0 / float64(stats.Cardinality)
		}
	}

	// For GIN indexes, assume lower selectivity for text search
	if idx.Type == stor.IndexTypeGIN {
		return 0.01 // Very selective for text search
	}

	return stats.Selectivity
}

// estimateIndexCost estimates the cost of using an index
func (qo *QueryOptimizer) estimateIndexCost(idx *stor.Index, _ []Filter, estimatedRows int) float64 {
	stats := qo.stats.IndexStats[idx.Name]
	if stats == nil {
		stats = &IndexStats{AvgBucketSize: 10.0}
	}

	// Cost model: index lookup cost + data retrieval cost
	indexLookupCost := 1.0 // Cost of index lookup
	dataRetrievalCost := float64(estimatedRows) * stats.AvgBucketSize

	return indexLookupCost + dataRetrievalCost
}

// UpdateStats updates collection statistics (called after inserts/updates/deletes)
func (qo *QueryOptimizer) UpdateStats(documentCount int, indexStats map[string]*IndexStats) {
	qo.stats.TotalDocuments = documentCount
	qo.stats.IndexStats = indexStats
}

// GetStats returns current collection statistics
func (qo *QueryOptimizer) GetStats() *CollectionStats {
	return qo.stats
}

File: ts/src/components/query/optimizer.ts
import { Query, Filter } from './knirvql';
import { Index, IndexType } from '../storage/index';

// QueryPlan represents an optimized execution plan for a query
export interface QueryPlan {
  // Execution strategy
  useIndex: boolean;
  indexName: string;
  indexType: IndexType;
  scanType: ScanType;

  // Query parameters
  filters: Filter[];
  limit: number;

  // Cost estimation
  estimatedCost: number;
  estimatedRows: number;

  // Execution details
  indexFilters: Filter[]; // Filters that can be pushed to index
  postFilters: Filter[]; // Filters that need post-processing
}

// ScanType represents the type of scan to perform
export enum ScanType {
  FullScan,
  IndexScan,
  IndexOnlyScan,
}

// CollectionStats holds statistics about a collection for cost estimation
export interface CollectionStats {
  totalDocuments: number;
  indexStats: Map<string, IndexStats>;
}

// IndexStats holds statistics about an index
export interface IndexStats {
  cardinality: number; // Number of unique values
  selectivity: number; // Fraction of documents matching a filter (0-1)
  avgBucketSize: number; // Average number of documents per index key
}

// QueryOptimizer analyzes queries and creates optimal execution plans
export class QueryOptimizer {
  private collection: string;
  private indexes: Index[];
  private stats: CollectionStats;

  constructor(collection: string, indexes: Index[], stats?: CollectionStats) {
    this.collection = collection;
    this.indexes = indexes;
    this.stats = stats || {
      totalDocuments: 1000, // Default estimate
      indexStats: new Map(),
    };

    // Initialize default stats for indexes
    for (const idx of this.indexes) {
      if (!this.stats.indexStats.has(idx.name)) {
        this.stats.indexStats.set(idx.name, {
          cardinality: 100,
          selectivity: 0.1,
          avgBucketSize: 10.0,
        });
      }
    }
  }

  // Optimize creates an optimal query plan for the given query
  optimize(query: Query): QueryPlan {
    const plan: QueryPlan = {
      useIndex: false,
      indexName: '',
      indexType: IndexType.BTree,
      scanType: ScanType.FullScan,
      filters: query.filters || [],
      limit: query.limit || 0,
      estimatedCost: 0,
      estimatedRows: this.stats.totalDocuments,
      indexFilters: [],
      postFilters: [],
    };

    // If no filters, always do full scan
    if (!query.filters || query.filters.length === 0) {
      plan.estimatedRows = (query.limit && query.limit > 0) ? Math.min(this.stats.totalDocuments, query.limit) : this.stats.totalDocuments;
      plan.estimatedCost = plan.estimatedRows;
      return plan;
    }

    // Analyze filters and find best index
    const bestIndex = this.selectBestIndex(query.filters);

    if (bestIndex) {
      plan.useIndex = true;
      plan.indexName = bestIndex.index.name;
      plan.indexType = bestIndex.index.type;
      plan.indexFilters = bestIndex.indexFilters;
      plan.postFilters = bestIndex.postFilters;

      // Estimate cost and rows for index scan
      plan.estimatedRows = this.estimateIndexRows(bestIndex.index, bestIndex.indexFilters);
      plan.estimatedCost = this.estimateIndexCost(bestIndex.index, bestIndex.indexFilters, plan.estimatedRows);

      // Determine scan type
      if (bestIndex.postFilters.length === 0) {
        plan.scanType = ScanType.IndexOnlyScan;
      } else {
        plan.scanType = ScanType.IndexScan;
      }
    } else {
      // Full scan
      plan.estimatedCost = this.stats.totalDocuments;
      plan.postFilters = query.filters;
    }

    // Apply limit to estimated rows if specified
    if (plan.limit > 0 && plan.estimatedRows > plan.limit) {
      plan.estimatedRows = plan.limit;
    }

    return plan;
  }

  // selectBestIndex chooses the best index for the given filters
  private selectBestIndex(filters: Filter[]): { index: Index; indexFilters: Filter[]; postFilters: Filter[] } | null {
    if (filters.length === 0) {
      return null;
    }

    type IndexCandidate = {
      index: Index;
      indexFilters: Filter[];
      postFilters: Filter[];
      cost: number;
      rows: number;
    };

    const candidates: IndexCandidate[] = [];

    // Evaluate each index
    for (const idx of this.indexes) {
      const { indexFilters, postFilters } = this.analyzeIndexSuitability(idx, filters);

      if (indexFilters.length > 0) {
        const estimatedRows = this.estimateIndexRows(idx, indexFilters);
        const estimatedCost = this.estimateIndexCost(idx, indexFilters, estimatedRows);

        candidates.push({
          index: idx,
          indexFilters,
          postFilters,
          cost: estimatedCost,
          rows: estimatedRows,
        });
      }
    }

    if (candidates.length === 0) {
      return null;
    }

    // Sort by cost (lowest first)
    candidates.sort((a, b) => a.cost - b.cost);

    const best = candidates[0];
    return {
      index: best.index,
      indexFilters: best.indexFilters,
      postFilters: best.postFilters,
    };
  }

  // analyzeIndexSuitability determines which filters can use an index
  private analyzeIndexSuitability(idx: Index, filters: Filter[]): { indexFilters: Filter[]; postFilters: Filter[] } {
    const indexFilters: Filter[] = [];
    const postFilters: Filter[] = [];

    for (const filter of filters) {
      if (this.canUseIndex(idx, filter)) {
        indexFilters.push(filter);
      } else {
        postFilters.push(filter);
      }
    }

    return { indexFilters, postFilters };
  }

  // canUseIndex checks if a filter can use the given index
  private canUseIndex(idx: Index, filter: Filter): boolean {
    switch (idx.type) {
      case IndexType.BTree:
        return this.canUseBTreeIndex(idx, filter);
      case IndexType.GIN:
        return this.canUseGINIndex(idx, filter);
      case IndexType.HNSW:
        return false; // Not for regular filters
      default:
        return false;
    }
  }

  // canUseBTreeIndex checks if a B-Tree index can be used for a filter
  private canUseBTreeIndex(idx: Index, filter: Filter): boolean {
    // Check if the filter field is indexed
    return idx.fields.includes(filter.key);
  }

  // canUseGINIndex checks if a GIN index can be used for a filter
  private canUseGINIndex(_idx: Index, _filter: Filter): boolean {
    // GIN indexes can be used for text search
    // For now, support CONTAINS operations
    return true; // Simplified - assume GIN can handle most text filters
  }

  // estimateIndexRows estimates the number of rows returned by an index scan
  private estimateIndexRows(idx: Index, filters: Filter[]): number {
    if (filters.length === 0) {
      return this.stats.totalDocuments;
    }

    const stats = this.stats.indexStats.get(idx.name);
    if (!stats) {
      return Math.floor(this.stats.totalDocuments * 0.1);
    }

    // Start with total documents
    let estimatedRows = this.stats.totalDocuments;

    // Apply selectivity for each filter
    for (const filter of filters) {
      const selectivity = this.estimateFilterSelectivity(idx, filter);
      estimatedRows *= selectivity;
    }

    return Math.max(1, Math.floor(estimatedRows));
  }

  // estimateFilterSelectivity estimates the selectivity of a filter
  private estimateFilterSelectivity(idx: Index, _filter: Filter): number {
    const stats = this.stats.indexStats.get(idx.name);
    if (!stats) {
      return 0.1; // Default selectivity
    }

    // For B-Tree indexes, use cardinality
    if (idx.type === IndexType.BTree) {
      if (stats.cardinality > 0) {
        return 1.0 / stats.cardinality;
      }
    }

    // For GIN indexes, assume lower selectivity for text search
    if (idx.type === IndexType.GIN) {
      return 0.01; // Very selective for text search
    }

    return stats.selectivity;
  }

  // estimateIndexCost estimates the cost of using an index
  private estimateIndexCost(idx: Index, _filters: Filter[], estimatedRows: number): number {
    const stats = this.stats.indexStats.get(idx.name);
    const avgBucketSize = stats ? stats.avgBucketSize : 10.0;

    // Cost model: index lookup cost + data retrieval cost
    const indexLookupCost = 1.0;
    const dataRetrievalCost = estimatedRows * avgBucketSize;

    return indexLookupCost + dataRetrievalCost;
  }

  // UpdateStats updates collection statistics (called after inserts/updates/deletes)
  updateStats(documentCount: number, indexStats: Map<string, IndexStats>): void {
    this.stats.totalDocuments = documentCount;
    this.stats.indexStats = indexStats;
  }

  // GetStats returns current collection statistics
  getStats(): CollectionStats {
    return this.stats;
  }
}